### ISSUE TITLE
viewPagerAndroid not working after push and back

### ISSUE THREAD

**Issue by hezheop** [wix/react-native-navigation#1317]
I use react-native-tab-view , and push to a view ,and back to the last view , the viewPagerAndroid's touch is not working , and the Tab Button clicks without effect 
Android:7.0,5.0
react-native : 0.44,
react-native-navigation:1.1.80

**Comment by Ehesp** [comment]
Is the onPress triggering JS side?

**Comment by enahum** [comment]
I'm having the same issue, whenever I push a view and then come back no `onPress` work in any component rendered inside the ViewPagerAndroid, and swiping the pager becomes really really slow

**Comment by hezheop** [comment]
That is  the ViewPagerAndroid 's bug I think , and I solve that by using react-native-tab-view ' s  TabViewPagerPan   [https://github.com/react-native-community/react-native-tab-view](url)

**Comment by enahum** [comment]
still happens with the suggested library or any other, including https://github.com/leecade/react-native-swiper and the funny thing is that it only happens on Android, IOS works just fine

**Comment by kelvinpompey** [comment]
I am also experiencing this issue.

**Comment by enahum** [comment]
So the way I solved it is that on android instead of doing a push and pop I did a show and dismiss modal

**Comment by PARAGJYOTI** [comment]
Its a bug in ViewPagerAndroid . In ios its works fine as its based on ScrollView .Any library that implements ViewPagerAndroid ,same problem will occur . Using Modal instead of push will work , but not a good solution.

**Comment by dzunglht** [comment]
this is a critical bug towards android app using viewpager, please help support this ticket guys
Tons of thanks

**Comment by t2n3qqq** [comment]
@dzunglht   You can try something like this:
```javascript
import React, { PureComponent } from 'react';
import {
  View,
  Dimensions,
} from 'react-native';

export default class ViewPagerWrapper extends PureComponent {

  constructor(props) {
    super(props);
    this.props.navigator && 
      this.props.navigator.setOnNavigatorEvent(this.onNavigatorEvent.bind(this));
  }

  onNavigatorEvent (event) {
    switch(event.id) {
      case 'willAppear':
        this._reattach();
        break;
    }
  }

  state = {
    width: Dimensions.get('window').width,
    height: Dimensions.get('window').height,
  }

  render() {
    return (
      <View style={[this.props.style, { flex: 1 }]}>
        <View style={{ width: this.state.width, height: this.state.height }}>
          {this.props.children}
        </View>
      </View>
    )
  }

  _y = 0.5

  _reattach = () => {
    this.setState({
      height: this.state.height - this._y,
    }, () => {
      this._y = this._y === 0.5 ? -0.5 : 0.5;
    });
  }
```
and then use:
```javascript
<ViewPagerWrapper navigator={this.props.navigator}>
  <YourAwesomeComponent>
    //Component that uses ViewPagerAndroid
    ...
  </YourAwesomeComponent>
</ViewPagerWrapper>
```
Maybe this is not most elegant solution, but it works for me.

**Comment by meow0703** [comment]
@guyca Please have an official fix asap, this is such a critical bug, above way doesn't work.
Thanks

**Comment by jykun** [comment]
look at this https://github.com/jykun/react-native-swiper.git, i change viewPageAndroid to scrollView.

**Comment by guyca** [comment]
@meow0703 @dzunglht Can you guys upload an example where the bug is reproducible? I doubt this is a bug in RNN, it's probably an issue in the actual component which will require a PR to RN.

**Comment by mannol** [comment]
I just came here to confirm that this happens to me too.

**Comment by mannol** [comment]
@guyca Steps to reproduce:
  - create root with startTabBasedApp() with `x` tabs
  - in one tab, render `ViewPagerAndroid`
  - from that tab push some screen with this.props.navigator.push()
  - press back on the newly pushed screen
  - the `ViewPagerAndroid` no longer works

**Comment by dzunglht** [comment]
Dear folks,
8 days passed, we're looking for your official fix asap 
Many thanks

**Comment by PARAGJYOTI** [comment]
Dear wix team , can you make a alternative viewPager component to
ViewPagerAndroid . ViewpagerAndroid has only few features that doesnot
include pagePadding or snapped carousel like features . I know
react-native-navigation has a built in Viewpager class used for TopTabs
scrolling . So please create a top lavel Viewpager Component .

@guyca please , think about it.

**Comment by meow0703** [comment]
@all @guyca Please have a look on below elegant library (built for react-native in mind) and refer how efficiently they use their viewPager, we should extend and/or make use of their effort. Hope that help.
Ref: https://github.com/appintheair/react-native-looped-carousel

**Comment by dzunglht** [comment]
@guyca where are you now. We're waiting for your official fix.

**Comment by Ehesp** [comment]
If you guys are waiting on this internally or something then I'd try and look at alternative options for now, there is no timeline on when this will be fixed.

**Comment by guyca** [comment]
Hey guys

I managed to take some time today to look into this issue. I think this is actually a bug in the native `ViewPager`. It seems to be [tracking its layout state](https://android.googlesource.com/platform/frameworks/support/+/jb-dev/v4/java/android/support/v4/view/ViewPager.java#1157) when the view is attached. Since the view is already laid out when it's added again to the screen - onLayout isn't called which is where `mIsFirstLayout` is set to false.
Because the `ViewPager` can't smooth scroll back to the selected item before `onLayout` stage is completed (since it doesn't know the item size) - smooth scrolling doesn't work.

We can easily trick the system into measuring and going through the entire layout cycle of the ViewPager (essentially invalidating the view) by changing the dimensions of the `ViewPagerAndroid` when the visibility of the screen changes.

```js
export default class ViewPagerAndroidScreen extends React.Component {
constructor(props) {
    super(props);
    this.props.navigator.setOnNavigatorEvent(this.onNavigatorEvent.bind(this));
    this.state = {
      visible: true
    }
  }

  onNavigatorEvent(event) {
    if (event.id === 'willAppear') {
      this.setState({
        visible: true
      });
    }
    if (event.id === 'willDisappear') {
      this.setState({
        visible: false
      });
    }
  }

  render() {
    return (
      <ViewPagerAndroid
        style={[styles.viewPager, {flex: this.state.visible ? 1 : 0}]}>
        <View>
          <Text>First page</Text>
        </View>
        <View>
          <Text>Second page</Text>
        </View>
      </ViewPagerAndroid>
    );
  }
}
```

**Comment by PARAGJYOTI** [comment]
Thanx @guyca  . Finally . What a relief .

**Comment by andrerfneves** [comment]
The above solution didn't work for me. I went ahead and used a Modal and it worked flawlessly.

**Comment by guyca** [comment]
@andrerfneves Modal is displayed over current screen - so the ViewPager isn't detached from window (which is the root cause of this bug)

**Comment by andrerfneves** [comment]
@guyca I'm aware of the reason why Modals work. I'm stating that the solution you provided above didn't work for me, aka re-render/calculating height/width again didn't solve it.

**Comment by mannol** [comment]
I'm here to confirm that @guyca's solution works. Thanks!

**Comment by kelvinpompey** [comment]
I am using https://github.com/archriss/react-native-snap-carousel as an alternative and it is working pretty well.

**Comment by mannol** [comment]
@kelvinpompey the reason for using viewPagerAndroid (or any control based on it) is that RefreshControl does not work with paged ScrollView

**Comment by meow0703** [comment]
@guyca when we have the official fix in next release ? Newbiews don't have time to search out this ticket
Thanks

**Comment by drpiou** [comment]
...annnnnnnd it is still unpatched. I have same issue

**Comment by rodribech20** [comment]
I have same issue. Is there an official fix?  Temporally I am using the forks mentioned above.

**Comment by rodolfobarretoweb** [comment]
follow

**Comment by dzunglht** [comment]
@here @all @guyca Dear the responsible author, please resolve the ticket issue and put it into the release roadmap guys, please resolve the ticket, not try finding a workaround !
Tons of thanks

**Comment by andrerfneves** [comment]
@dzunglht this is an open source initiative. Software development is what it is today because people like @guyca (and companies like Wix) and others spend countless hours devoting their energy and knowledge to create these complex solutions so we can build better applications.

Don't demand that things get fixed.

Fix it yourself, or use the workaround pointed above.

**Comment by heron2014** [comment]
Hi @kelvinpompey I am using ```react-native-snap-carousel``` but whenever I import Carousel I got this error: 

<img width="285" alt="screen shot 2017-10-21 at 18 40 32" src="https://user-images.githubusercontent.com/8476884/31854646-96a80a28-b694-11e7-9dae-43abd0ce43cd.png">

Have you had this issue by any chance or anyone?
Is there any workaround it? 

Thanks

**Comment by a648132694** [comment]
@t2n3qqq Thank you , it also works great for me

**Comment by iksent** [comment]
guyca commented on 12 Jul -That Solution **works** for me for "react-native-tab-view".
I use {flex: this.state.visible ? 1 : 0} for container of TabViewAnimated.

So i have:

 <View style={{flex: this.state.visible ? 1 : 0}}>
                <TabViewAnimated ... />
</View>

**Comment by smooJitter** [comment]
@heron2014 did you get is to work

**Comment by heron2014** [comment]
@smooJitter yes, I have upgraded RN to 0.48.0 which I was due anyway and it started to work without any workaround :)

**Comment by Annihil** [comment]
I confirm @guyca solution fixes the issue with [React-Native-ViewPager](https://github.com/zbtang/React-Native-ViewPager) on Android, kudos

**Comment by janiokq** [comment]
![image](https://user-images.githubusercontent.com/10770735/36833012-fa025dae-1d68-11e8-87c4-15025d2ab7f7.png)
![image](https://user-images.githubusercontent.com/10770735/36833101-3e06a05a-1d69-11e8-94b2-486024d5eb49.png)


In the push
RelativeLayout.removeView( previous view  )
In the pop
RelativeLayout.addview(previous view)

RelativeLayout   extends ViewGroup  

In the native
ViewPager is removed by ViewGroup and added to

Will there be this bug

**Comment by janiokq** [comment]
You can

![image](https://user-images.githubusercontent.com/10770735/36834226-27e3b41c-1d6d-11e8-825b-568a00e1ffdc.png)
Function is the
 private void pushScreenToVisibleStack(LayoutParams layoutParams,
                                          final Screen nextScreen,
                                          final Screen previousScreen,
                                          @Nullable final Promise onPushComplete,
                                          @Nullable final Screen.OnDisplayListener onDisplay, final Boolean runanime) 

![image](https://user-images.githubusercontent.com/10770735/36834249-416c97a0-1d6d-11e8-95af-9edec41279f7.png)
Function is the
public void pop(final boolean animated, final double jsPopTimestamp, @Nullable final OnScreenPop onScreenPop) 

![image](https://user-images.githubusercontent.com/10770735/36834275-551c39ea-1d6d-11e8-94c0-8f9afdd7881d.png)
Function is the
private void swapScreens(boolean animated, final Screen toRemove, Screen previous, OnScreenPop onScreenPop) 

All changes in
com.reactnativenavigation.screens ->ScreenStack.java

The problem will be repaired

But when multiple pages exist, the graphics memory will be high

Because all pages are displayed on the page



Do anyone have a better solution? I'm a JavaScript code farm native code that isn't very well understood.

**Comment by janiokq** [comment]
I found the real reason

In the native

react-native-navigation
The implementation is to add and remove views in a ViewGroup

I saw the implementation of ViewPager
![image](https://user-images.githubusercontent.com/10770735/36883094-c83e537c-1e12-11e8-9b3d-2650f417ea91.png)

mFirstLayout  It's this guy
This state is used to record whether the viewpager is the first initialization of the layout

He has such a piece of code
![image](https://user-images.githubusercontent.com/10770735/36883569-3f125a36-1e16-11e8-9bf1-7d8824cf81be.png)

When navigation is switched on the page
Viewpager is removed from the view and added
It triggered this event
mFirstLayout = true;

Viewpager can't roll.
You can do this

![image](https://user-images.githubusercontent.com/10770735/36883648-ca4807ea-1e16-11e8-9177-1aac2347152a.png)
Using this class to use viewpager

There's no problem.

**Comment by Annihil** [comment]
@janiokq How did you use a custom class for your ViewPager, please explain how to do

**Comment by gusgoose** [comment]
@janiokq Good work finding the problem - is this a solution that works for you? And if so, please could you explain how we can do the same?

**Comment by ashokkumar88** [comment]
@t2n3qqq this solutin works for me. But if you have multiple carousel/ViewpagerAndroid in a screen then instead of `setOnNavigatorEvent` use `addOnNavigatorEvent`

`this.props.navigator.addOnNavigatorEvent(this.onNavigatorEvent.bind(this));`

**Comment by paulbargaoanu** [comment]
In case anyone is still encountering this issue while using react-native-tab-view, I would suggest using <PagerScroll /> instead of <PagerAndroid />, unless you really need <PagerAndroid /> for some reason. I am developing an app for both iOS and Android and initially chose for <PagerScroll /> for iOS and <PagerAndroid /> for Android and encountered this unexpected bug only on Android. After some ressearch I thought I'd just try changing the pager and thus far it seems to work flawlessly. Hope this might help someone!

**Comment by jkvim** [comment]
@janiokq  I had try your snippet and wrap my custom ViewPager, but not work. could you explain how to do, thanks.

**Comment by edgarbonillag** [comment]
Hi everyone,

The library `'react-native-tab-view'` uses React Native's `ViewPagerAndroid` as default pager for Android and this component is not working well. 

I've found an easy SOLUTION that made my app run without any problem. I just had to tell the `'react-native-tab-view'` to use the `PagerScroll` (which is the default for iOS and its based on RN's `ScrollView`) and that's it. To do that, you just `import {PagerScroll} from 'react-native-tab-view'` pass the following `prop` to the main `TabView` component:

```
<TabView
   ...
   renderPager={(props) => <PagerScroll {...props}/>}
   ...
/>
```

**Comment by rat-moonshine** [comment]
Hi! Can someone guide me how to pass react-native-navigation screens (`Navigation.registerComponent(Screens.ScreenA, () => require('./ScreenA'));`) to use with react native tab vie? I'm having having error 'invariant violation: native module can not be null'. Thanks!

```
const renderScene = SceneMap({
    first: Screens.ScreenA,
    second: Screens.ScreenB
  });

....

render() {
    return (
      <View>
        <TabView
            navigationState={{ index, routes }}
            renderScene={renderScene}
            onIndexChange={setIndex}
            initialLayout={initialLayout}
          />
     </View>
    );
  }
```

**Comment by yedidyak** [comment]
Closing this as part of a maintainer cleanup of open issues older than one year.

Decision: Closed: stale/non-critical; no evidence it is a critical current-version RNN issue.

If this still reproduces on the latest React Native Navigation with a minimal reproduction, please open a new issue with current RN/RNN versions and repro steps.
